File: ruby/lib/pactman/nonprofit_check_plus/transport.rb
# frozen_string_literal: true

module Pactman
  module NonprofitCheckPlus
    # Hooks the test suite substitutes for real time. Not part of the public API
    # and not covered by semantic versioning; do not depend on it.
    #
    # @api private
    TransportHooks = Data.define(:sleep, :random, :monotonic) do
      def initialize(sleep: nil, random: nil, monotonic: nil)
        super
      end
    end

    # HTTP transport: authentication headers, timeouts, retries with jittered
    # backoff, `Retry-After` handling, client-side throttling, and mapping
    # responses to the error taxonomy.
    #
    # The API key lives in a closure, not an instance variable. It is written
    # into the `Authorization` header at send time and is never reachable from
    # `inspect`, `pp`, `to_yaml`, `Marshal`, an error, or any diagnostic output.
    #
    # @api private
    class Transport
      # A parsed HTTP response plus the metadata callers need.
      Response = Data.define(:status, :request_id, :body, :attempts)

      # Headers the SDK owns. A default or per-request header of the same name,
      # in any case, is replaced rather than sent alongside.
      RESERVED_HEADERS = %w[accept user-agent authorization content-type].freeze

      REQUEST_ID_HEADERS = %w[x-request-id x-correlation-id request-id].freeze

      # @param api_key [String]
      # @param config [ResolvedConfig]
      # @param hooks [TransportHooks, nil]
      def initialize(api_key, config, hooks = nil)
        @authorization = self.class.send(:credential, api_key)
        @config = config
        @sleep = hooks&.sleep || ->(seconds) { Kernel.sleep(seconds) }
        @random = hooks&.random || -> { Kernel.rand }
        @monotonic = hooks&.monotonic || -> { Process.clock_gettime(Process::CLOCK_MONOTONIC) }
        @throttle_lock = Mutex.new
        @next_request_at = nil
      end

      # Sends one logical request, retrying under the policy in force.
      #
      # @return [Response]
      # @raise [ApiError, TimeoutError, NetworkError]
      def perform(http_method:, path:, body: nil, timeout: nil, retry_option: nil, headers: nil)
        policy = Config.resolve_retry(@config.retry_policy, retry_option)
        request = build_request(http_method, path, body, timeout, headers)
        attempt = 0

        loop do
          attempt += 1
          throttle

          response, error, cause = attempt_once(request, attempt)

          if response.nil?
            raise error, cause: cause if attempt > policy.max_retries

            wait_before_retry(attempt, policy, nil)
            next
          end

          outcome = read_response(response, attempt)
          return outcome if outcome.is_a?(Response)

          raise outcome if attempt > policy.max_retries || !policy.retryable_status?(outcome.status)

          wait_before_retry(attempt, policy, outcome.retry_after_seconds)
        end
      end

      def inspect
        "#<#{self.class.name} #{@config.base_url}>"
      end

      # Delay before the next attempt, in seconds.
      #
      # A valid `Retry-After` wins outright. Otherwise the delay grows
      # exponentially from `initial_delay`, is capped at `max_delay`, and — with
      # jitter on — is randomized across the whole range so concurrent clients
      # spread out. Rounded to the millisecond.
      #
      # @param attempt [Integer] the attempt that just failed, starting at 1.
      # @param policy [RetryPolicy]
      # @param retry_after_seconds [Numeric, nil]
      # @param random [#call] returns a Float in `[0, 1)`.
      # @return [Float]
      def self.compute_retry_delay(attempt, policy, retry_after_seconds, random = -> { Kernel.rand })
        if policy.respect_retry_after && retry_after_seconds && retry_after_seconds >= 0
          return retry_after_seconds.to_f.round(3)
        end

        exponential = policy.initial_delay * (policy.backoff_factor**(attempt - 1))
        capped = [exponential, policy.max_delay].min

        (policy.jitter ? random.call * capped : capped).to_f.round(3)
      end

      # Reads `Retry-After` as either a delay in seconds or an HTTP date.
      #
      # @param headers [Hash{String => String}] lowercase names.
      # @param now [Time]
      # @return [Numeric, nil] seconds, or `nil` when absent or unreadable.
      def self.read_retry_after(headers, now = Time.now)
        raw = headers["retry-after"]
        return nil if raw.nil?

        trimmed = raw.strip
        return nil if trimmed.empty?

        seconds = Integer(trimmed, 10, exception: false) || Float(trimmed, exception: false)
        return seconds if seconds&.finite? && seconds >= 0

        date = parse_http_date(trimmed)
        date && [0, date - now].max
      end

      # Normalizes the envelope's `errors` into a list.
      #
      # @return [Array<ApiErrorDetail>]
      def self.normalize_api_errors(errors)
        case errors
        when Array then errors.grep(Hash).map { |entry| ApiErrorDetail.new(entry) }
        when Hash then [ApiErrorDetail.new(errors)]
        when String then errors.strip.empty? ? [] : [ApiErrorDetail.new({ "reason" => errors })]
        else []
        end
      end

      def self.parse_http_date(value)
        Time.httpdate(value)
      rescue ArgumentError
        begin
          Time.rfc2822(value)
        rescue ArgumentError
          nil
        end
      end
      private_class_method :parse_http_date

      def self.credential(api_key)
        header = "Bearer #{api_key}".freeze

        -> { header }
      end
      private_class_method :credential

      private

      def build_request(http_method, path, body, timeout, headers)
        Http::Request.new(
          http_method: http_method,
          url: "#{@config.base_url}#{path}",
          headers: build_headers(Config.validate_headers(headers, "headers"), !body.nil?),
          body: body.nil? ? nil : JSON.generate(body),
          timeout: timeout.nil? ? @config.timeout : Config.validate_timeout(timeout)
        )
      end

      # The parsed success, or the {ApiError} a non-2xx response maps to.
      def read_response(response, attempt)
        status = response.status.to_i
        headers = normalize_headers(response.headers)
        parsed = parse_body(response.body)
        request_id = read_request_id(headers)

        return Response.new(status:, request_id:, body: parsed, attempts: attempt) if (200..299).cover?(status)

        api_error(status, parsed, request_id, self.class.read_retry_after(headers), attempt)
      end

      def build_headers(per_request, has_body)
        headers = {}

        [@config.default_headers, per_request].each do |source|
          source.each { |name, value| headers[name.downcase] = value }
        end

        RESERVED_HEADERS.each { |name| headers.delete(name) }

        headers["accept"] = "application/json"
        headers["user-agent"] = @config.user_agent
        headers["authorization"] = @authorization.call
        headers["content-type"] = "application/json" if has_body
        headers
      end

      # One attempt. Returns `[response]`, or `[nil, error, cause]` when no
      # response arrived. Only `StandardError` is rescued: anything else is the
      # caller cancelling, and it is theirs.
      def attempt_once(request, attempt)
        [@config.http_adapter.call(request)]
      rescue ::Timeout::Error => e
        [nil, TimeoutError.new("The request timed out after #{format("%g", request.timeout)}s.",
                               timeout: request.timeout, attempts: attempt), e]
      rescue StandardError => e
        [nil, NetworkError.new("The request to the Pactman API failed: #{e.message}", attempts: attempt), e]
      end

      # Spaces requests when `max_requests_per_second` is configured.
      def throttle
        limit = @config.max_requests_per_second
        return if limit.nil?

        wait = @throttle_lock.synchronize do
          now = @monotonic.call
          scheduled_at = @next_request_at.nil? ? now : [now, @next_request_at].max
          @next_request_at = scheduled_at + (1.0 / limit)
          scheduled_at - now
        end

        @sleep.call(wait) if wait.positive?
      end

      def wait_before_retry(attempt, policy, retry_after_seconds)
        @sleep.call(self.class.compute_retry_delay(attempt, policy, retry_after_seconds, @random))
      end

      def normalize_headers(headers)
        (headers || {}).to_h do |name, value|
          [name.to_s.downcase, value.is_a?(Array) ? value.join(", ") : value.to_s]
        end
      end

      def read_request_id(headers)
        REQUEST_ID_HEADERS.lazy.filter_map { |name| headers[name] }.first
      end

      # The parsed JSON, the text itself when it is not JSON, or `nil` when there
      # was no body. An unparseable body is still evidence, so it is kept.
      def parse_body(body)
        text = body.to_s.dup.force_encoding(Encoding::UTF_8)
        text = text.scrub unless text.valid_encoding?
        return nil if text.strip.empty?

        begin
          JSON.parse(text)
        rescue JSON::ParserError
          text
        end
      end

      def api_error(status, parsed, request_id, retry_after, attempts)
        envelope = parsed.is_a?(Hash) ? parsed : nil
        api_errors = self.class.normalize_api_errors(envelope&.fetch("errors", nil))

        ApiError.for_status(
          status: status,
          api_message: api_message(api_errors, envelope, parsed),
          api_code: envelope && envelope["code"].is_a?(Numeric) ? envelope["code"] : nil,
          api_errors: api_errors,
          request_id: request_id,
          retry_after_seconds: retry_after,
          raw: parsed,
          attempts: attempts
        )
      end

      def api_message(api_errors, envelope, parsed)
        reasons = api_errors.map(&:reason).select { |reason| reason.is_a?(String) && !reason.strip.empty? }

        return reasons.join("; ") unless reasons.empty?
        return envelope["message"] if envelope && envelope["message"].is_a?(String)

        parsed.strip[0, 500] if parsed.is_a?(String) && !parsed.strip.empty?
      end
    end
  end
end

File: java/sdk/src/main/java/org/pactman/nonprofitcheckplus/http/Transport.java
package org.pactman.nonprofitcheckplus.http;

import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.Instant;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import org.pactman.nonprofitcheckplus.config.ClientConfig;
import org.pactman.nonprofitcheckplus.config.RequestOptions;
import org.pactman.nonprofitcheckplus.config.RetryOptions;
import org.pactman.nonprofitcheckplus.exceptions.ApiErrorInit;
import org.pactman.nonprofitcheckplus.exceptions.PactmanApiException;
import org.pactman.nonprofitcheckplus.exceptions.PactmanConfigurationException;
import org.pactman.nonprofitcheckplus.exceptions.PactmanException;
import org.pactman.nonprofitcheckplus.exceptions.PactmanNetworkException;
import org.pactman.nonprofitcheckplus.exceptions.PactmanTimeoutException;
import org.pactman.nonprofitcheckplus.internal.Json;
import org.pactman.nonprofitcheckplus.internal.JsonException;
import org.pactman.nonprofitcheckplus.models.ApiErrorDetail;
import org.pactman.nonprofitcheckplus.models.ResponseBody;

/**
 * HTTP transport: authentication headers, timeouts, cancellation, retries with
 * jittered backoff, {@code Retry-After} handling, and mapping responses to the
 * exception taxonomy.
 *
 * <p>The API key is written into the {@code Authorization} header at send time
 * and is never stored on a request record, an exception, or any diagnostic
 * output.
 */
public final class Transport {

    private final String apiKey;
    private final ClientConfig config;
    private final TransportHooks hooks;

    /** Guards {@link #nextRequestAtMs}; a client is shared across threads. */
    private final Object throttleLock = new Object();

    private long nextRequestAtMs;

    /**
     * Creates the transport.
     *
     * @param apiKey the validated API key.
     * @param config the resolved configuration.
     * @param hooks  the clock and random source, or {@code null} for the real ones.
     */
    public Transport(String apiKey, ClientConfig config, TransportHooks hooks) {
        this.apiKey = apiKey;
        this.config = config;
        this.hooks = hooks == null ? TransportHooks.DEFAULT : hooks;
    }

    /**
     * Sends a request, retrying under the resolved policy.
     *
     * @param method  {@code GET} or {@code POST}.
     * @param path    the path, already encoded, to append to the base URL.
     * @param body    the request body to serialize as JSON, or {@code null} for none.
     * @param options per-request overrides, or {@code null}.
     * @return the parsed response.
     * @throws PactmanApiException     when the API returns a status this SDK
     *                                 surfaces as an error.
     * @throws PactmanTimeoutException when an attempt exceeds the timeout and no
     *                                 retries remain.
     * @throws PactmanNetworkException when the request produces no response, or
     *                                 the caller cancels it.
     */
    public TransportResponse send(
            String method, String path, Object body, RequestOptions options) {
        RetryOptions retry = options != null && options.retry() != null
                ? options.retry()
                : config.retry();
        long timeoutMs = options != null && options.timeoutMs() != null
                ? options.timeoutMs()
                : config.timeoutMs();
        URI url = URI.create(config.baseUrl() + path);
        String payload = body == null ? null : Json.write(body);
        Map<String, String> headers = buildHeaders(options, payload != null);

        int attempt = 0;

        for (;;) {
            attempt += 1;
            throwIfCancelled(0);
            throttle();

            Attempt outcome = attempt(url, method, headers, payload, timeoutMs, attempt);

            if (outcome.response != null) {
                HttpResponse<String> response = outcome.response;
                Object parsed = parseBody(response.body());
                String requestId = readRequestId(response);

                if (response.statusCode() >= 200 && response.statusCode() < 300) {
                    return new TransportResponse(
                            response.statusCode(),
                            requestId,
                            new ResponseBody(asFields(parsed)),
                            attempt);
                }

                Double retryAfterSeconds = readRetryAfter(
                        response.headers().firstValue("retry-after").orElse(null), Instant.now());

                PactmanApiException apiError = PactmanApiException.fromStatus(
                        buildApiErrorInit(
                                response.statusCode(), parsed, requestId, retryAfterSeconds, attempt));

                if (attempt > retry.maxRetries()
                        || !retry.isRetryableStatus(response.statusCode())) {
                    throw apiError;
                }

                waitBeforeRetry(attempt, retry, retryAfterSeconds);
                continue;
            }

            // A timeout or a transport failure.
            if (attempt > retry.maxRetries() || outcome.cancelled) {
                throw outcome.error;
            }

            waitBeforeRetry(attempt, retry, null);
        }
    }

    /** Header names this SDK owns; a caller's value for one of these is dropped. */
    private static final List<String> RESERVED_HEADERS =
            java.util.Arrays.asList("authorization", "accept", "user-agent", "content-type");

    private Map<String, String> buildHeaders(RequestOptions options, boolean hasBody) {
        Map<String, String> supplied = new LinkedHashMap<>(config.defaultHeaders());

        if (options != null) {
            supplied.putAll(options.headers());
        }

        Map<String, String> headers = new LinkedHashMap<>();

        for (Map.Entry<String, String> header : supplied.entrySet()) {
            // Dropped rather than overwritten: HTTP header names are
            // case-insensitive and the JDK's builder appends, so a caller's
            // `authorization` would otherwise ride along beside the real
            // credential instead of losing to it.
            if (!RESERVED_HEADERS.contains(
                    header.getKey().toLowerCase(java.util.Locale.ROOT))) {
                headers.put(header.getKey(), header.getValue());
            }
        }

        headers.put("Accept", "application/json");
        headers.put("User-Agent", config.userAgent());

        if (hasBody) {
            headers.put("Content-Type", "application/json");
        }

        return headers;
    }

    private Attempt attempt(
            URI url,
            String method,
            Map<String, String> headers,
            String payload,
            long timeoutMs,
            int attemptNumber) {
        HttpRequest request = buildRequest(url, method, headers, payload, timeoutMs);
        CompletableFuture<HttpResponse<String>> pending = config.httpExchange().sendAsync(request);

        try {
            return Attempt.of(pending.get(timeoutMs, TimeUnit.MILLISECONDS));
        } catch (TimeoutException expired) {
            pending.cancel(true);

            return Attempt.failed(
                    new PactmanTimeoutException(
                            "The request timed out after " + timeoutMs + "ms.",
                            timeoutMs,
                            attemptNumber,
                            expired),
                    false);
        } catch (InterruptedException cancelled) {
            pending.cancel(true);
            // Restore the flag: the caller asked to stop, and code above this
            // frame needs to see that too.
            Thread.currentThread().interrupt();

            return Attempt.failed(
                    new PactmanNetworkException(
                            "The request was cancelled by the caller.", attemptNumber, cancelled),
                    true);
        } catch (ExecutionException failed) {
            Throwable cause = unwrap(failed);

            if (cause instanceof java.net.http.HttpTimeoutException) {
                return Attempt.failed(
                        new PactmanTimeoutException(
                                "The request timed out after " + timeoutMs + "ms.",
                                timeoutMs,
                                attemptNumber,
                                cause),
                        false);
            }

            return Attempt.failed(
                    new PactmanNetworkException(
                            "The request to the Pactman API failed: " + describe(cause),
                            attemptNumber,
                            cause),
                    false);
        }
    }

    private HttpRequest buildRequest(
            URI url, String method, Map<String, String> headers, String payload, long timeoutMs) {
        HttpRequest.Builder request = HttpRequest.newBuilder(url)
                // Belt and braces with the future timeout below: this one tears
                // the exchange down at the socket rather than leaving it running
                // behind a future nobody is waiting on any more.
                .timeout(Duration.ofMillis(timeoutMs));

        for (Map.Entry<String, String> header : headers.entrySet()) {
            try {
                // setHeader, not header: the latter appends, which would send a
                // duplicate whenever a caller spells a header in another case.
                request.setHeader(header.getKey(), header.getValue());
            } catch (IllegalArgumentException restricted) {
                throw new PactmanConfigurationException(
                        "The header \"" + header.getKey()
                                + "\" cannot be set on an outgoing request: "
                                + restricted.getMessage());
            }
        }

        request.setHeader("Authorization", "Bearer " + apiKey);

        if (payload == null) {
            request.method(method, HttpRequest.BodyPublishers.noBody());
        } else {
            request.method(
                    method, HttpRequest.BodyPublishers.ofString(payload, StandardCharsets.UTF_8));
        }

        return request.build();
    }

    /** Spaces requests when {@code maxRequestsPerSecond} is configured. */
    private void throttle() {
        Double limit = config.maxRequestsPerSecond();

        if (limit == null) {
            return;
        }

        long waitMs;

        synchronized (throttleLock) {
            long intervalMs = Math.max(1L, Math.round(1000.0d / limit));
            long now = System.currentTimeMillis();
            long scheduledAt = Math.max(now, nextRequestAtMs);
            nextRequestAtMs = scheduledAt + intervalMs;
            waitMs = scheduledAt - now;
        }

        if (waitMs > 0) {
            sleep(waitMs, 0);
        }
    }

    private void waitBeforeRetry(int attempt, RetryOptions retry, Double retryAfterSeconds) {
        sleep(computeRetryDelay(attempt, retry, retryAfterSeconds, hooks.random()), attempt);
        throwIfCancelled(attempt);
    }

    private void sleep(long millis, int attempt) {
        try {
            hooks.sleep(millis);
        } catch (InterruptedException cancelled) {
            Thread.currentThread().interrupt();

            throw new PactmanNetworkException(
                    "The request was cancelled by the caller.", attempt, cancelled);
        }
    }

    private static void throwIfCancelled(int attempt) {
        if (Thread.currentThread().isInterrupted()) {
            throw new PactmanNetworkException(
                    "The request was cancelled by the caller.", attempt, null);
        }
    }

    /**
     * The delay before the next attempt.
     *
     * <p>A valid {@code Retry-After} wins outright. Otherwise the delay grows
     * exponentially from {@code initialDelayMs}, is capped at {@code maxDelayMs},
     * and — with jitter on — is randomized across the whole range so concurrent
     * clients spread out.
     *
     * @param attempt           the attempt that just failed, counting from one.
     * @param retry             the policy in force.
     * @param retryAfterSeconds the server's {@code Retry-After}, or {@code null}.
     * @param random            a value in {@code [0, 1)} for jitter.
     * @return the delay in milliseconds.
     */
    public static long computeRetryDelay(
            int attempt, RetryOptions retry, Double retryAfterSeconds, double random) {
        if (retry.respectRetryAfter() && retryAfterSeconds != null && retryAfterSeconds >= 0) {
            return Math.round(retryAfterSeconds * 1000.0d);
        }

        double exponential =
                retry.initialDelayMs() * Math.pow(retry.backoffFactor(), attempt - 1.0d);
        double capped = Math.min(exponential, retry.maxDelayMs());

        return Math.round(retry.jitter() ? random * capped : capped);
    }

    /**
     * Reads {@code Retry-After} as either a delay in seconds or an HTTP date.
     *
     * @param raw the header value, or {@code null} when the server sent none.
     * @param now the instant to measure an HTTP date against.
     * @return the delay in seconds, or {@code null} when there is no usable value.
     */
    public static Double readRetryAfter(String raw, Instant now) {
        if (raw == null) {
            return null;
        }

        String trimmed = raw.trim();

        if (trimmed.isEmpty()) {
            return null;
        }

        try {
            double seconds = Double.parseDouble(trimmed);

            // "Infinity" and "1e400" parse, and no wait can honour either.
            if (Double.isFinite(seconds) && seconds >= 0) {
                return seconds;
            }

            return null;
        } catch (NumberFormatException notANumber) {
            // Falls through to the HTTP-date form below.
        }

        try {
            ZonedDateTime when =
                    ZonedDateTime.parse(trimmed, DateTimeFormatter.RFC_1123_DATE_TIME);

            return Math.max(0.0d, (when.toInstant().toEpochMilli() - now.toEpochMilli()) / 1000.0d);
        } catch (DateTimeParseException notADate) {
            return null;
        }
    }

    private static ApiErrorInit buildApiErrorInit(
            int status, Object parsed, String requestId, Double retryAfterSeconds, int attempts) {
        ResponseBody envelope = parsed instanceof Map ? new ResponseBody(asFields(parsed)) : null;
        List<ApiErrorDetail> apiErrors = envelope == null
                ? java.util.Collections.<ApiErrorDetail>emptyList()
                : envelope.getErrors();

        List<String> reasons = new ArrayList<>();

        for (ApiErrorDetail detail : apiErrors) {
            String reason = detail.getReason();

            if (reason != null && !reason.trim().isEmpty()) {
                reasons.add(reason);
            }
        }

        String apiMessage;

        if (!reasons.isEmpty()) {
            apiMessage = String.join("; ", reasons);
        } else if (envelope != null && envelope.getMessage() != null) {
            apiMessage = envelope.getMessage();
        } else if (parsed instanceof String && !((String) parsed).trim().isEmpty()) {
            String text = ((String) parsed).trim();
            apiMessage = text.length() > 500 ? text.substring(0, 500) : text;
        } else {
            apiMessage = null;
        }

        return ApiErrorInit.builder(status)
                .apiMessage(apiMessage)
                .apiCode(envelope == null ? null : envelope.getCode())
                .apiErrors(apiErrors)
                .requestId(requestId)
                .retryAfterSeconds(retryAfterSeconds)
                .raw(parsed)
                .attempts(attempts)
                .build();
    }

    /**
     * Decodes a response body, keeping whatever the server sent when it is not
     * JSON — an unparseable body is still evidence.
     */
    private static Object parseBody(String text) {
        if (text == null || text.trim().isEmpty()) {
            return null;
        }

        try {
            return Json.parse(text);
        } catch (JsonException notJson) {
            return text;
        }
    }

    private static String readRequestId(HttpResponse<String> response) {
        for (String header : new String[] {"x-request-id", "x-correlation-id", "request-id"}) {
            String value = response.headers().firstValue(header).orElse(null);

            if (value != null) {
                return value;
            }
        }

        return null;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> asFields(Object value) {
        return value instanceof Map
                ? (Map<String, Object>) value
                : java.util.Collections.<String, Object>emptyMap();
    }

    private static Throwable unwrap(Throwable error) {
        Throwable cause = error.getCause();

        while (cause instanceof CompletionException && cause.getCause() != null) {
            cause = cause.getCause();
        }

        return cause == null ? error : cause;
    }

    private static String describe(Throwable cause) {
        String message = cause.getMessage();

        return message == null || message.isEmpty() ? cause.getClass().getSimpleName() : message;
    }

    /** One attempt's outcome: a response, or the failure that replaced it. */
    private static final class Attempt {
        private final HttpResponse<String> response;
        private final PactmanException error;
        private final boolean cancelled;

        private Attempt(HttpResponse<String> response, PactmanException error, boolean cancelled) {
            this.response = response;
            this.error = error;
            this.cancelled = cancelled;
        }

        static Attempt of(HttpResponse<String> response) {
            return new Attempt(response, null, false);
        }

        static Attempt failed(PactmanException error, boolean cancelled) {
            return new Attempt(null, error, cancelled);
        }
    }
}
